File: CodingInterviews/java/42.maximum-continuous-subarray-sum.java
class SolutionA {
    public int maxSubArray(int[] nums) {
        if(nums == null || nums.length <= 0){
            throw new IllegalArgumentException("Empty array.");
        }
        int maxSum = nums[0];
        int sum = 0;
        for(int num:nums){
            sum += num;
            maxSum = Math.max(maxSum, sum);
            if(sum < 0){
                sum = 0;
            }
        }
        return maxSum;
    }
}

class SolutionB {
    public int maxSubArray(int[] nums) {
        if(nums == null || nums.length <= 0){
            throw new IllegalArgumentException("Empty array.");
        }
        int maxSum = nums[0];
        int[] dp = new int[nums.length+1];
        for(int i = 1;i<=nums.length;++i){
            dp[i] = Math.max(0, dp[i-1]) + nums[i-1];
            maxSum = Math.max(dp[i], maxSum);
        }
        return maxSum;
    }
}

File: coding-interviews/cpp/42.maximum-continuous-subarray-sum.cpp
#include<iostream>
#include<vector>

using namespace std;

class Solution {
public:
    int FindGreatestSumOfSubArray(vector<int> array) {
        int sum = 0;
        int maxSum = std::numeric_limits<int>::min();
        for(int num:array){
            if(sum>=0){
                sum += num;
            }else{
                sum = num;
            }
            maxSum = max(sum,maxSum);
        }
        return maxSum;
    }
};

int main(){
    Solution s;
    assert(s.FindGreatestSumOfSubArray({6,-3,-2,7,-15,1,2,2}) == 8);
    assert(s.FindGreatestSumOfSubArray({1,-2,3,10,-4,7,2,-5}) == 18);
    assert(s.FindGreatestSumOfSubArray({-2,-8,-1,-5,-9}) == -1);
}